# Crisis notebook

Trying to access wikidata information for historical crisis.
Currently looking at wars.

## Test case
### The Great Northern War

name: label english
description: description english

entity_id	id
2019	    P276    # location id
288	        P582    # end time id
21423	    Q151616 # The Great Northern War id
177	        P580    # start time id
84	        P31     # instance of id

## Table info

tables:
table_name
wd_aliases
wd_claim_globecoordinate
wd_claim_monolingualtext
wd_claim_no_value
wd_claim_quantity
wd_claim_some_value
wd_claim_string
wd_claim_time
wd_claim_wikibase_entityid
wd_datatypes
wd_descriptions
wd_entities
wd_labels
wd_qualifier_globecoordinate
wd_qualifier_monolingualtext
wd_qualifier_no_value
wd_qualifier_quantity
wd_qualifier_some_value
wd_qualifier_string
wd_qualifier_time
wd_qualifier_wikibase_entityid
wd_reference_globecoordinate
wd_reference_monolingualtext
wd_reference_no_value
wd_reference_quantity
wd_reference_some_value
wd_reference_string
wd_reference_time
wd_reference_wikibase_entityid
wd_sitelink_badges
wd_sitelinks

tables and columns:
table_name	column_name	data_type
wd_aliases	entity_id	BIGINT
wd_aliases	language	VARCHAR
wd_aliases	alias	VARCHAR
wd_claim_globecoordinate	claim_id	BIGINT
wd_claim_globecoordinate	rank	VARCHAR
wd_claim_globecoordinate	entity_id	BIGINT
wd_claim_globecoordinate	property_id	BIGINT
wd_claim_globecoordinate	datatype	VARCHAR
wd_claim_globecoordinate	latitude	DOUBLE
wd_claim_globecoordinate	longitude	DOUBLE
wd_claim_globecoordinate	precision	DOUBLE
wd_claim_globecoordinate	globe_entity_id	BIGINT
wd_claim_monolingualtext	claim_id	BIGINT
wd_claim_monolingualtext	rank	VARCHAR
wd_claim_monolingualtext	entity_id	BIGINT
wd_claim_monolingualtext	property_id	BIGINT
wd_claim_monolingualtext	datatype	VARCHAR
wd_claim_monolingualtext	language	VARCHAR
wd_claim_monolingualtext	text	VARCHAR
wd_claim_no_value	claim_id	BIGINT
wd_claim_no_value	rank	VARCHAR
wd_claim_no_value	entity_id	BIGINT
wd_claim_no_value	property_id	BIGINT
wd_claim_no_value	datatype	VARCHAR
wd_claim_quantity	claim_id	BIGINT
wd_claim_quantity	rank	VARCHAR
wd_claim_quantity	entity_id	BIGINT
wd_claim_quantity	property_id	BIGINT
wd_claim_quantity	datatype	VARCHAR
wd_claim_quantity	amount	DOUBLE
wd_claim_quantity	lower_bound	DOUBLE
wd_claim_quantity	upper_bound	DOUBLE
wd_claim_quantity	unit_entity_id	BIGINT
wd_claim_some_value	claim_id	BIGINT
wd_claim_some_value	rank	VARCHAR
wd_claim_some_value	entity_id	BIGINT
wd_claim_some_value	property_id	BIGINT
wd_claim_some_value	datatype	VARCHAR
wd_claim_string	claim_id	BIGINT
wd_claim_string	rank	VARCHAR
wd_claim_string	entity_id	BIGINT
wd_claim_string	property_id	BIGINT
wd_claim_string	datatype	VARCHAR
wd_claim_string	value	VARCHAR
wd_claim_time	claim_id	BIGINT
wd_claim_time	rank	VARCHAR
wd_claim_time	entity_id	BIGINT
wd_claim_time	property_id	BIGINT
wd_claim_time	datatype	VARCHAR
wd_claim_time	time	VARCHAR
wd_claim_time	timezone	INTEGER
wd_claim_time	before	INTEGER
wd_claim_time	after	INTEGER
wd_claim_time	precision	INTEGER
wd_claim_time	calendarmodel_entity_id	BIGINT
wd_claim_wikibase_entityid	claim_id	BIGINT
wd_claim_wikibase_entityid	rank	VARCHAR
wd_claim_wikibase_entityid	entity_id	BIGINT
wd_claim_wikibase_entityid	property_id	BIGINT
wd_claim_wikibase_entityid	datatype	VARCHAR
wd_claim_wikibase_entityid	value_entity_id	BIGINT
wd_datatypes	entity_id	BIGINT
wd_datatypes	datatype	VARCHAR
wd_descriptions	entity_id	BIGINT
wd_descriptions	language	VARCHAR
wd_descriptions	description	VARCHAR
wd_entities	entity_id	BIGINT
wd_entities	id	VARCHAR
wd_labels	entity_id	BIGINT
wd_labels	language	VARCHAR
wd_labels	label	VARCHAR
wd_qualifier_globecoordinate	order	INTEGER
wd_qualifier_globecoordinate	claim_id	BIGINT
wd_qualifier_globecoordinate	property_id	BIGINT
wd_qualifier_globecoordinate	datatype	VARCHAR
wd_qualifier_globecoordinate	latitude	DOUBLE
wd_qualifier_globecoordinate	longitude	DOUBLE
wd_qualifier_globecoordinate	precision	DOUBLE
wd_qualifier_globecoordinate	globe_entity_id	BIGINT
wd_qualifier_monolingualtext	order	INTEGER
wd_qualifier_monolingualtext	claim_id	BIGINT
wd_qualifier_monolingualtext	property_id	BIGINT
wd_qualifier_monolingualtext	datatype	VARCHAR
wd_qualifier_monolingualtext	language	VARCHAR
wd_qualifier_monolingualtext	text	VARCHAR
wd_qualifier_no_value	order	INTEGER
wd_qualifier_no_value	claim_id	BIGINT
wd_qualifier_no_value	property_id	BIGINT
wd_qualifier_no_value	datatype	VARCHAR
wd_qualifier_quantity	order	INTEGER
wd_qualifier_quantity	claim_id	BIGINT
wd_qualifier_quantity	property_id	BIGINT
wd_qualifier_quantity	datatype	VARCHAR
wd_qualifier_quantity	amount	DOUBLE
wd_qualifier_quantity	lower_bound	DOUBLE
wd_qualifier_quantity	upper_bound	DOUBLE
wd_qualifier_quantity	unit_entity_id	BIGINT
wd_qualifier_some_value	order	INTEGER
wd_qualifier_some_value	claim_id	BIGINT
wd_qualifier_some_value	property_id	BIGINT
wd_qualifier_some_value	datatype	VARCHAR
wd_qualifier_string	order	INTEGER
wd_qualifier_string	claim_id	BIGINT
wd_qualifier_string	property_id	BIGINT
wd_qualifier_string	datatype	VARCHAR
wd_qualifier_string	value	VARCHAR
wd_qualifier_time	order	INTEGER
wd_qualifier_time	claim_id	BIGINT
wd_qualifier_time	property_id	BIGINT
wd_qualifier_time	datatype	VARCHAR
wd_qualifier_time	time	VARCHAR
wd_qualifier_time	timezone	INTEGER
wd_qualifier_time	before	INTEGER
wd_qualifier_time	after	INTEGER
wd_qualifier_time	precision	INTEGER
wd_qualifier_time	calendarmodel_entity_id	BIGINT
wd_qualifier_wikibase_entityid	order	INTEGER
wd_qualifier_wikibase_entityid	claim_id	BIGINT
wd_qualifier_wikibase_entityid	property_id	BIGINT
wd_qualifier_wikibase_entityid	datatype	VARCHAR
wd_qualifier_wikibase_entityid	value_entity_id	BIGINT
wd_reference_globecoordinate	order	INTEGER
wd_reference_globecoordinate	claim_id	BIGINT
wd_reference_globecoordinate	property_id	BIGINT
wd_reference_globecoordinate	datatype	VARCHAR
wd_reference_globecoordinate	latitude	DOUBLE
wd_reference_globecoordinate	longitude	DOUBLE
wd_reference_globecoordinate	precision	DOUBLE
wd_reference_globecoordinate	globe_entity_id	BIGINT
wd_reference_monolingualtext	order	INTEGER
wd_reference_monolingualtext	claim_id	BIGINT
wd_reference_monolingualtext	property_id	BIGINT
wd_reference_monolingualtext	datatype	VARCHAR
wd_reference_monolingualtext	language	VARCHAR
wd_reference_monolingualtext	text	VARCHAR
wd_reference_no_value	order	INTEGER
wd_reference_no_value	claim_id	BIGINT
wd_reference_no_value	property_id	BIGINT
wd_reference_no_value	datatype	VARCHAR
wd_reference_quantity	order	INTEGER
wd_reference_quantity	claim_id	BIGINT
wd_reference_quantity	property_id	BIGINT
wd_reference_quantity	datatype	VARCHAR
wd_reference_quantity	amount	DOUBLE
wd_reference_quantity	lower_bound	DOUBLE
wd_reference_quantity	upper_bound	DOUBLE
wd_reference_quantity	unit_entity_id	BIGINT
wd_reference_some_value	order	INTEGER
wd_reference_some_value	claim_id	BIGINT
wd_reference_some_value	property_id	BIGINT
wd_reference_some_value	datatype	VARCHAR
wd_reference_string	order	INTEGER
wd_reference_string	claim_id	BIGINT
wd_reference_string	property_id	BIGINT
wd_reference_string	datatype	VARCHAR
wd_reference_string	value	VARCHAR
wd_reference_time	order	INTEGER
wd_reference_time	claim_id	BIGINT
wd_reference_time	property_id	BIGINT
wd_reference_time	datatype	VARCHAR
wd_reference_time	time	VARCHAR
wd_reference_time	timezone	INTEGER
wd_reference_time	before	INTEGER
wd_reference_time	after	INTEGER
wd_reference_time	precision	INTEGER
wd_reference_time	calendarmodel_entity_id	BIGINT
wd_reference_wikibase_entityid	order	INTEGER
wd_reference_wikibase_entityid	claim_id	BIGINT
wd_reference_wikibase_entityid	property_id	BIGINT
wd_reference_wikibase_entityid	datatype	VARCHAR
wd_reference_wikibase_entityid	value_entity_id	BIGINT
wd_sitelink_badges	entity_id	BIGINT
wd_sitelink_badges	site	VARCHAR
wd_sitelink_badges	badge_entity_id	BIGINT
wd_sitelinks	entity_id	BIGINT
wd_sitelinks	site	VARCHAR
wd_sitelinks	title	VARCHAR

In [1]:
from sqlframe_gizmosql import GizmoSQLSession, GizmoSQLTable
import sqlframe_gizmosql.functions as F

from hereutil import here
import yaml

with open(here("db_secret.yaml"), "r") as f:
    db_params = yaml.safe_load(f)


# Create an SQLFrame (PySpark) session connected to GizmoSQL
session: GizmoSQLSession = GizmoSQLSession.builder \
    .config("gizmosql.uri", db_params["uri"]) \
    .config("gizmosql.username", db_params["username"]) \
    .config("gizmosql.password", db_params["password"]) \
    .getOrCreate()

c = F.col

# -------------------------------------------------------------------
# Tables
# -------------------------------------------------------------------

wd_entities: GizmoSQLTable = session.table("books.wd_entities")
wd_labels: GizmoSQLTable = session.table("books.wd_labels")
wd_descriptions: GizmoSQLTable = session.table("books.wd_descriptions")

wd_claim_time: GizmoSQLTable = session.table("books.wd_claim_time")
wd_claim_wikibase_entityid: GizmoSQLTable = session.table(
    "books.wd_claim_wikibase_entityid"
)

# -------------------------------------------------------------------
# Constants
# -------------------------------------------------------------------

# Great Northern War
ENTITY_ID = 21423

# Properties
P31_INSTANCE_OF = 84
P580_START_TIME = 177
P582_END_TIME = 288
P276_LOCATION = 2019

# -------------------------------------------------------------------
# Event label + description (English)
# -------------------------------------------------------------------

event_metadata = (
    wd_entities.alias("e")
    .join(
        wd_labels.alias("l"),
        (
            (c("e.entity_id") == c("l.entity_id")) &
            (c("l.language") == F.lit("en"))
        ),
        how="left"
    )
    .join(
        wd_descriptions.alias("d"),
        (
            (c("e.entity_id") == c("d.entity_id")) &
            (c("d.language") == F.lit("en"))
        ),
        how="left"
    )
    .filter(c("e.entity_id") == ENTITY_ID)
    .select(
        c("e.id").alias("qid"),
        c("l.label").alias("label"),
        c("d.description").alias("description")
    )
)

event_metadata.show()

# -------------------------------------------------------------------
# Start + end dates
# -------------------------------------------------------------------

event_times = (
    wd_claim_time
    .filter(
        (c("entity_id") == ENTITY_ID) &
        (
            c("property_id").isin(
                P580_START_TIME,
                P582_END_TIME
            )
        )
    )
    .select(
        "property_id",
        "time",
        "precision",
        "rank"
    )
)

event_times.show()

# -------------------------------------------------------------------
# Instance of
# -------------------------------------------------------------------

instance_of = (
    wd_claim_wikibase_entityid.alias("claims")

    .join(
        wd_entities.alias("target"),
        c("claims.value_entity_id") == c("target.entity_id")
    )

    .join(
        wd_labels.alias("labels"),
        (
            (c("target.entity_id") == c("labels.entity_id")) &
            (c("labels.language") == F.lit("en"))
        ),
        how="left"
    )

    .filter(
        (c("claims.entity_id") == ENTITY_ID) &
        (c("claims.property_id") == P31_INSTANCE_OF)
    )

    .select(
        c("target.id").alias("target_qid"),
        c("labels.label").alias("instance_of")
    )
)

instance_of.show()

# -------------------------------------------------------------------
# Locations
# -------------------------------------------------------------------

locations = (
    wd_claim_wikibase_entityid.alias("claims")

    .join(
        wd_entities.alias("target"),
        c("claims.value_entity_id") == c("target.entity_id")
    )

    .join(
        wd_labels.alias("labels"),
        (
            (c("target.entity_id") == c("labels.entity_id")) &
            (c("labels.language") == F.lit("en"))
        ),
        how="left"
    )

    .filter(
        (c("claims.entity_id") == ENTITY_ID) &
        (c("claims.property_id") == P276_LOCATION)
    )

    .select(
        c("target.id").alias("location_qid"),
        c("labels.label").alias("location")
    )
)

locations.show()

OperationalError: CANCELLED: [FlightSQL] context canceled (Canceled; DoGet: endpoint 0: [])

In [2]:
from adbc_driver_flightsql import dbapi
from hereutil import here
import yaml

# -------------------------------------------------------------------
# Load DB credentials
# -------------------------------------------------------------------

with open(here("db_secret.yaml"), "r") as f:
    db_params = yaml.safe_load(f)

# -------------------------------------------------------------------
# Constants
# -------------------------------------------------------------------

# Great Northern War
ENTITY_ID = 21423

# Properties
P31_INSTANCE_OF = 84
P580_START_TIME = 177
P582_END_TIME = 288
P276_LOCATION = 2019

# -------------------------------------------------------------------
# Connect
# -------------------------------------------------------------------

with dbapi.connect(
    db_params["uri"],
    db_kwargs=dict(
        username=db_params["username"],
        password=db_params["password"]
    )
) as con:

    with con.cursor() as cur:

        # -----------------------------------------------------------
        # Event metadata (label + description)
        # -----------------------------------------------------------

        metadata_query = f"""
        SELECT
            e.id AS qid,
            l.label AS label,
            d.description AS description
        FROM books.wd_entities e

        LEFT JOIN books.wd_labels l
            ON e.entity_id = l.entity_id
           AND l.language = 'en'

        LEFT JOIN books.wd_descriptions d
            ON e.entity_id = d.entity_id
           AND d.language = 'en'

        WHERE e.entity_id = {ENTITY_ID}
        """

        print("\n=== EVENT METADATA ===")
        cur.execute(metadata_query)
        print(cur.fetch_arrow_table())

        # -----------------------------------------------------------
        # Start + end dates
        # -----------------------------------------------------------

        dates_query = f"""
        SELECT
            property_id,
            time,
            precision,
            rank
        FROM books.wd_claim_time
        WHERE entity_id = {ENTITY_ID}
          AND property_id IN ({P580_START_TIME}, {P582_END_TIME})
        """

        print("\n=== EVENT DATES ===")
        cur.execute(dates_query)
        print(cur.fetch_arrow_table())

        # -----------------------------------------------------------
        # Instance of
        # -----------------------------------------------------------

        instance_query = f"""
        SELECT
            target.id AS target_qid,
            labels.label AS instance_of

        FROM books.wd_claim_wikibase_entityid claims

        JOIN books.wd_entities target
            ON claims.value_entity_id = target.entity_id

        LEFT JOIN books.wd_labels labels
            ON target.entity_id = labels.entity_id
           AND labels.language = 'en'

        WHERE claims.entity_id = {ENTITY_ID}
          AND claims.property_id = {P31_INSTANCE_OF}
        """

        print("\n=== INSTANCE OF ===")
        cur.execute(instance_query)
        print(cur.fetch_arrow_table())

        # -----------------------------------------------------------
        # Locations
        # -----------------------------------------------------------

        location_query = f"""
        SELECT
            target.id AS location_qid,
            labels.label AS location

        FROM books.wd_claim_wikibase_entityid claims

        JOIN books.wd_entities target
            ON claims.value_entity_id = target.entity_id

        LEFT JOIN books.wd_labels labels
            ON target.entity_id = labels.entity_id
           AND labels.language = 'en'

        WHERE claims.entity_id = {ENTITY_ID}
          AND claims.property_id = {P276_LOCATION}
        """

        print("\n=== LOCATIONS ===")
        cur.execute(location_query)
        print(cur.fetch_arrow_table())


=== EVENT METADATA ===
pyarrow.Table
qid: string
label: string
description: string
----
qid: [["Q151616"]]
label: [["Great Northern War"]]
description: [["conflict between anti-Swedish coalition Led by Russia and Swedish empires in 1700–1721"]]

=== EVENT DATES ===
pyarrow.Table
property_id: int64
time: string
precision: int32
rank: string
----
property_id: [[177],[288]]
time: [["+1700-02-22T00:00:00Z"],["+1721-09-10T00:00:00Z"]]
precision: [[11],[11]]
rank: [["normal"],["normal"]]

=== INSTANCE OF ===
pyarrow.Table
target_qid: string
instance_of: string
----
target_qid: [["Q198"]]
instance_of: [["war"]]

=== LOCATIONS ===
pyarrow.Table
location_qid: string
location: string
----
location_qid: [["Q27468"]]
location: [["Eastern Europe"]]


took 13m 34s to generate

=== EVENT METADATA ===
pyarrow.Table
qid: string
label: string
description: string
----
qid: [["Q151616"]]
label: [["Great Northern War"]]
description: [["conflict between anti-Swedish coalition Led by Russia and Swedish empires in 1700–1721"]]

=== EVENT DATES ===
pyarrow.Table
property_id: int64
time: string
precision: int32
rank: string
----
property_id: [[177],[288]]
time: [["+1700-02-22T00:00:00Z"],["+1721-09-10T00:00:00Z"]]
precision: [[11],[11]]
rank: [["normal"],["normal"]]

=== INSTANCE OF ===
pyarrow.Table
...
location: string
----
location_qid: [["Q27468"]]
location: [["Eastern Europe"]]

In [3]:
from adbc_driver_flightsql import dbapi
from hereutil import here
import yaml

# -------------------------------------------------------------------
# Load DB credentials
# -------------------------------------------------------------------

with open(here("db_secret.yaml"), "r") as f:
    db_params = yaml.safe_load(f)

# -------------------------------------------------------------------
# Input war name
# -------------------------------------------------------------------

war_name = "Great Northern War"

# -------------------------------------------------------------------
# Connect
# -------------------------------------------------------------------

with dbapi.connect(
    db_params["uri"],
    db_kwargs=dict(
        username=db_params["username"],
        password=db_params["password"]
    )
) as con:

    with con.cursor() as cur:

        # -----------------------------------------------------------
        # Find matching Wikidata entities by English label
        # -----------------------------------------------------------

        query = f"""
        SELECT
            e.entity_id,
            e.id AS wikidata_id,
            l.label,
            d.description

        FROM books.wd_labels l

        JOIN books.wd_entities e
            ON l.entity_id = e.entity_id

        LEFT JOIN books.wd_descriptions d
            ON e.entity_id = d.entity_id
           AND d.language = 'en'

        WHERE l.language = 'en'
          AND lower(l.label) = lower('{war_name}')

        LIMIT 20
        """

        cur.execute(query)

        result = cur.fetch_arrow_table()

        print(result)

pyarrow.Table
entity_id: int64
wikidata_id: string
label: string
description: string
----
entity_id: [[21423]]
wikidata_id: [["Q151616"]]
label: [["Great Northern War"]]
description: [["conflict between anti-Swedish coalition Led by Russia and Swedish empires in 1700–1721"]]


took 20m 39.2s to finish

pyarrow.Table
entity_id: int64
wikidata_id: string
label: string
description: string
----
entity_id: [[21423]]
wikidata_id: [["Q151616"]]
label: [["Great Northern War"]]
description: [["conflict between anti-Swedish coalition Led by Russia and Swedish empires in 1700–1721"]]